examples/aligned_accessor/aligned_accessor.cpp

```diff
@@ -0,0 +1,734 @@
+  // We have coverage for the Windows case (at least with MSVC) above,
+  // so we can resort to the POSIX function posix_memalign.
+  const int err = posix_memalign(&ptr, byte_alignment, num_bytes);
+  if(err != 0) {
+    if(err == EINVAL) {
+      throw std::runtime_error("posix_memalign failed: alignment not a power of two");
+    } else if(err == ENOMEM) {
+      throw std::runtime_error("posix_memalign failed: insufficient memory");
+    } else {
+      throw std::runtime_error("posix_memalign failed: unknown error");
+    }
+  }
+#else
+  ptr = std::aligned_alloc(byte_alignment, num_bytes);
+#endif
+
+  return {reinterpret_cast<ElementType*>(ptr), deleter};
+}
+
+template<class ElementType, std::size_t byte_alignment>
+struct aligned_array_allocation {
+  static_assert(byte_alignment >= sizeof(ElementType),
+		"byte_alignment must be at least sizeof(ElementType).");
+  static constexpr std::size_t overalignment = byte_alignment / sizeof(ElementType);
+  static_assert(overalignment * sizeof(ElementType) == byte_alignment,
+		"overalignment * sizeof(ElementType) must equal byte_alignment.");
+
+  allocation_t<ElementType> allocation{nullptr, delete_raw<ElementType>{}};
+  aligned_pointer_t<ElementType, byte_alignment> aligned_pointer = nullptr;
+  std::size_t num_elements = 0;
+};
+
+template<class ElementType, std::size_t byte_alignment>
+aligned_array_allocation<ElementType, byte_alignment>
+allocate_aligned(std::size_t num_elements)
+{
+  if(num_elements == 0) {
+    return {};
+  }
+  auto allocation = allocate_raw<ElementType, byte_alignment>(num_elements);
+  aligned_pointer_t<ElementType, byte_alignment> aligned_pointer =
+    _MDSPAN_ASSUME_ALIGNED( ElementType, allocation.get(), byte_alignment );
+  return {std::move(allocation), aligned_pointer, num_elements};
+}
+
+// A dynamically allocated array of ElementType,
+// whose zeroth element has byte alignment byte_alignment.
+//
+// ElementType: A trivially copyable type whose size is a power of two bytes.
+//
+// byte_alignment: A power of two that is a multiple of sizeof(ElementType).
+template<class ElementType, std::size_t byte_alignment>
+class aligned_array {
+public:
+  aligned_array(std::size_t num_elements) :
+    alloc(allocate_aligned<ElementType, byte_alignment>(num_elements))
+  {}
+
+  // A pointer to the zeroth element of the array.
+  // It is aligned to byte_alignment bytes.
+  // All num_elements (constructor parameter) elements are valid.
+  aligned_pointer_t<ElementType, byte_alignment>
+  data_aligned() const noexcept
+  {
+    return _MDSPAN_ASSUME_ALIGNED( ElementType, alloc.aligned_pointer, byte_alignment );
+  }
+
+  // A pointer to the first (NOT zeroth) element of the array.
+  // It is aligned only to sizeof(ElementType) bytes.
+  // All num_elements (constructor parameter) elements are valid.
+  //
+  // This exists so that we can compare performance
+  // of aligned and unaligned access.
+  ElementType* data_unaligned() const noexcept
```

Comment:
Not sure if this is what we should do here instead of returning also address to element 0. After all you are now measuring something different, like the number of cachelines/pages touched by each thread might be different. I'd rather like to see that the attribute in and of itself does something.

Reply:
@crtrott My concern there was that the compiler has complete code visibility, so it could deduce that the allocation is aligned.  For the benchmark, I wanted to be able to give it an allocation that I _know_ is aligned to no more than `sizeof(ElementType)`.

Reply:
@crtrott I like the tag idea for the benchmarks, though!  I was a bit worried about possibly needing to do type conversion of x, y, and z on every invocation of the inner loop, but I think I could restructure the code to avoid that.

Reply:
@crtrott I tried out the tag idea, but I don't think it saves much code.  Each benchmark invocation still needs to name the const and nonconst array types and the tag types explicitly.

Perhaps y'all are missing the point a bit -- this is an _example_.  It's code for users to read, that teaches them how to write a custom accessor.  It's something more like a tutorial.  Reducing code duplication isn't really the point here; the point is simple, readable code that users could easily copy and paste.